File: quiz/deno/src/quiz/quiz.ts
import { User } from '@src/user.ts';
import { dataDir } from '@src/utils.ts';
import { Question } from '@src/quiz/question.ts';
import {
	Any,
	array,
	Decoverto,
	map,
	MapShape,
	model,
	property,
} from '@decoverto';
import { crypto } from 'https://deno.land/std/crypto/mod.ts';

@model()
export class Quiz {
	@property()
	id: string;

	@property()
	userId: string;

	@property()
	name: string;

	@property({
		toInstance: (value: string | undefined) => value,
		toPlain: (value: string | undefined) =>
			value === '' ? undefined : value,
	})
	description?: string;

	@property(array(() => Question))
	questions: Question<unknown>[] = [];

	@property()
	shareCode: string;

	@property(
		map(
			() => String,
			map(() => String, () => Boolean, {
				shape: MapShape.Object,
			}),
			{ shape: MapShape.Object },
		),
	)
	private responses: Map<string, Map<string, boolean>>;

	private constructor(
		id: string,
		userId: string,
		name: string,
		shareCode?: string,
		responses?: Map<string, Map<string, boolean>>,
	) {
		this.id = id;
		this.userId = userId;
		this.name = name;
		this.shareCode = shareCode ?? this.getShareCode();
		this.responses = responses ?? new Map();
	}

	public async delete(): Promise<Quiz[]> {
		const quizzes = await Quiz.getAllQuizzes();
		const index = quizzes.findIndex((quiz) => quiz.id === this.id);

		quizzes.splice(index, 1);
		Quiz.saveAllQuizzes(quizzes);
		return quizzes;
	}

	public addQuestion(question: Question<unknown>): Question<unknown> {
		this.questions.push(question);
		return question;
	}

	public removeQuestion(id: string): void {
		this.questions = this.questions.filter((question) =>
			question.id !== id
		);
	}

	public addResponse(answers: Map<string, boolean>): void {
		this.responses.set(crypto.randomUUID(), answers);
	}

	public getResponsesCount(): number {
		return this.responses.size;
	}

	public async save(): Promise<void> {
		const quizzes = await Quiz.getAllQuizzes();
		const index = quizzes.findIndex((quiz) => quiz.id === this.id);

		if (index === -1) {
			quizzes.push(this);
		} else {
			quizzes[index] = this;
		}
		Quiz.saveAllQuizzes(quizzes);
	}

	private getShareCode(): string {
		const encoder = new TextEncoder();
		const hashBuffer = crypto.subtle.digestSync(
			'SHA-256',
			encoder.encode(this.id),
		);
		const hashArray = Array.from(new Uint8Array(hashBuffer));
		const hashHex = hashArray.map((b) =>
			b.toString(16).padStart(2, '0')
		).join('').substring(0, 8);
		return parseInt(hashHex, 16).toString(36);
	}

	static isAnyQuizRegistered(user: User): Promise<boolean> {
		return this.getAllQuizzesForUser(user).then((quizzes) =>
			quizzes.length > 0
		);
	}

	private static async saveAllQuizzes(quizzes: Quiz[]): Promise<void> {
		const encoder = new TextEncoder();
		const decoverto = new Decoverto();

		const raw = decoverto.type(Quiz).instanceArrayToRaw(quizzes);
		await Deno.mkdir(dataDir, { recursive: true });

		await Deno.writeFile(
			`${dataDir}/quizzes.json`,
			encoder.encode(raw),
			{ create: true },
		);
	}

	private static async getAllQuizzes(): Promise<Quiz[]> {
		const decoder = new TextDecoder();
		const decoverto = new Decoverto();

		try {
			const content = await Deno.readFile(
				`${dataDir}/quizzes.json`,
			);
			const quizzes = decoverto.type(Quiz).rawToInstanceArray(
				decoder.decode(content),
			);
			return quizzes;
		} catch (_error) {
			return [];
		}
	}

	public static async create(user: User, name: string): Promise<Quiz> {
		const quizzes = await this.getAllQuizzes();
		const quiz = new Quiz(crypto.randomUUID(), user.id, name);

		quizzes.push(quiz);
		this.saveAllQuizzes(quizzes);
		return quiz;
	}

	public static async getAllQuizzesForUser(user: User): Promise<Quiz[]> {
		return (await this.getAllQuizzes()).filter((quiz) =>
			quiz.userId === user.id
		);
	}

	public static async getQuizById(user: User, id: string): Promise<Quiz> {
		const quizzes = await this.getAllQuizzes();
		const quiz = quizzes.find((quiz) =>
			quiz.id === id && quiz.userId === user.id
		);

		if (!quiz) {
			throw new Error(`Acces Violation`);
		}

		return quiz;
	}

	public static async getQuizByShareCode(
		shareCode: string,
	): Promise<Quiz | undefined> {
		const quizzes = await this.getAllQuizzes();
		const quiz = quizzes.find((quiz) =>
			quiz.shareCode === shareCode
		);
		return quiz;
	}

	public static async availableShareCodes(): Promise<string[]> {
		const quizzes = await this.getAllQuizzes();
		return quizzes.map((quiz) => quiz.shareCode);
	}
}

File: quiz/go/quiz/quiz.go
package quiz

import (
	"encoding/json"
	"errors"

	"github.com/DoktorLenz/kp-wise23-24/quiz/go/utils"
	"github.com/google/uuid"
)

type RawQuiz struct {
	ID          string                     `json:"id"`
	UserID      string                     `json:"userId"`
	Name        string                     `json:"name"`
	Description string                     `json:"description"`
	ShareCode   string                     `json:"shareCode"`
	Questions   []interface{}              `json:"questions"`
	Responses   map[string]map[string]bool `json:"responses"`
}

type Quiz struct {
	ID          string                     `json:"id"`
	UserID      string                     `json:"userId"`
	Name        string                     `json:"name"`
	Description string                     `json:"description"`
	ShareCode   string                     `json:"shareCode"`
	Questions   []IQuestion                `json:"questions"`
	Responses   map[string]map[string]bool `json:"responses"`
}

func (quiz *Quiz) Delete() ([]*Quiz, error) {
	quizzes, err := GetAllQuizzes()
	if err != nil {
		return nil, err
	}
	index := -1
	for i, q := range quizzes {
		if q.ID == quiz.ID {
			index = i
			break
		}
	}

	if index == -1 {
		return nil, errors.New("Quiz not found: " + quiz.ID)
	}

	quizzes = append(quizzes[:index], quizzes[index+1:]...)
	err = SaveAllQuizzes(quizzes)

	return quizzes, err

}

func (quiz *Quiz) Save() error {
	quizzes, err := GetAllQuizzes()
	if err != nil {
		return err
	}

	index := -1
	for i, q := range quizzes {
		if q.ID == quiz.ID {
			index = i
			break
		}
	}

	if index == -1 {
		return errors.New("Quiz not found: " + quiz.ID)
	}

	quizzes[index] = quiz

	err = SaveAllQuizzes(quizzes)

	return err
}

func Create(userId string, name string) (*Quiz, error) {
	quizzes, err := GetAllQuizzes()
	if err != nil {
		return nil, err
	}

	quiz := &Quiz{
		ID:        uuid.New().String(),
		UserID:    userId,
		Name:      name,
		Questions: make([]IQuestion, 0),
		Responses: make(map[string]map[string]bool),
	}

	quizzes = append(quizzes, quiz)

	err = SaveAllQuizzes(quizzes)

	return quiz, err
}

func GetAllQuizzesForUser(userID string) ([]*Quiz, error) {
	quizzes, err := GetAllQuizzes()
	if err != nil {
		return nil, err
	}

	var userQuizzes []*Quiz
	for _, quiz := range quizzes {
		if quiz.UserID == userID {
			userQuizzes = append(userQuizzes, quiz)
		}
	}

	return userQuizzes, nil
}

func GetAllQuizzes() ([]*Quiz, error) {
	var rawQuizzes []*RawQuiz
	fileContent, err := utils.ReadFile("quizzes.json")
	if err != nil || len(fileContent) == 0 {
		fileContent = []byte("[]")
	}
	err = json.Unmarshal(fileContent, &rawQuizzes)
	if err != nil {
		return nil, err
	}

	var quizzes []*Quiz

	for _, rawQuiz := range rawQuizzes {
		quiz := &Quiz{
			ID:          rawQuiz.ID,
			UserID:      rawQuiz.UserID,
			Name:        rawQuiz.Name,
			Description: rawQuiz.Description,
			ShareCode:   rawQuiz.ShareCode,
			Questions:   make([]IQuestion, 0),
			Responses:   rawQuiz.Responses,
		}
		for _, question := range rawQuiz.Questions {
			questionMap := question.(map[string]interface{})
			switch questionMap["__type"].(string) {
			case "ToggleQuestion":
				var tq *ToggleQuestion = &ToggleQuestion{}
				questionData, err := json.Marshal(question)
				if err != nil {
					return nil, err
				}
				err = json.Unmarshal(questionData, tq)
				if err != nil {
					return nil, err
				}
				quiz.Questions = append(quiz.Questions, tq)
			}
		}
		quizzes = append(quizzes, quiz)
	}

	return quizzes, nil
}

func SaveAllQuizzes(quizzes []*Quiz) error {
	data, err := json.Marshal(quizzes)
	if err != nil {
		return err
	}
	return utils.WriteFile("quizzes.json", data)
}
